README
        Emacs-wget README  ($Date: 2004/10/19 07:10:33 $)

This package contains an interface program for GNU wget on Emacs.

1. Introduction

   GNU Wget is a free software package for retrieving files using HTTP,
   HTTPS and FTP, the most widely-used Internet protocols.  It is a
   non-interactive command line tool.  More information, see wget web
   page:

   http://www.gnu.org/software/wget/

   GNU Wget is originally written by Hrvoje Niksic.


   Emacs-wget provides an interface of GNU wget for Emacs.  You can also
   call it from web browser on Emacs, like Emacs/W3 or emacs-w3m.

   The latest emacs-wget is available at:

   https://github.com/ataka/emacs-wget


2. Requirements

   GNU Emacs and GNU wget.

   Emacs: http://www.gnu.org/software/emacs/
   Wget:  http://www.gnu.org/software/wget/


3. Installation

   a.) make && make install

        $ tar xzvf emacs-wget-X.YY.tar.gz
        $ cd emacs-wget-X.YY
        $ make
        # make install

   a-1.) If make fails, put all *.el files into your load-path directory.


   b.) Put following expressions into your .emacs file.

        (autoload 'wget "wget" "wget interface for Emacs." t)
        (autoload 'wget-web-page "wget" "wget interface to download whole web page." t)


   c.) Setting for Web browser on Emacs:

   c-1.) With emacs-w3m, put the following code into your .emacs:

        (load "w3m-wget")

   c-2.) With Emacs/W3, put the following code into your .emacs:

        (autoload 'w3-wget "w3-wget" "wget interface for Emacs/W3." t)


   d) If wget version is 1.7 or less, put the following code into your .emacs:

        (setq wget-basic-options '("-v"))


   e) When you write your .wgetrc:

   e-1.) quiet = on

         Emacs-wget will fail to download.  Put the following into your .emacs:

        (setq wget-basic-options (cons "-equiet=off" wget-basic-options))

   e-2.) dir_prefix = PATH/TO/DOWNLOAD

         Variable wget-download-directory will be ignored.  Put the
         following into your .emacs:

        (setq wget-basic-options (cons "-P." wget-basic-options))

   e-3.) timestamping = on
         mirror = on  (mirror=on automatically sets timestamping=on)

         These options conflict with -nc option.  If your wget-default-options
         or wget-ftp-default-options have -nc option, get rid of it.

   e-4.) logfile = log

         *wget* buffer will appear but show no information, because the
         output from wget is robbed by the 'log' file.  I don't have
         good idea, except not displaying surd *wget* buffer:

        (setq wget-process-buffer nil)


   f) If the system can not search wget command, tell emacs where to
      find it.

        (setq wget-command "C:/cygwin/bin/wget")


4. Running with...

   (emacs-version)               wget -V
   ---------------               -------
    20.7 (Meadow 1.15)            1.8.2 (cygwin)
    21.3.50 (CVS)                 1.8.1, 1.8.2, 1.9.1

   If you find emacs-wget work with other environment above, please tell
   me <masayuki.ataka@gmail.com>.


5. Contact address

   E-mail to <masayuki.ataka@gmail.com>.
   Any information, bug-report, request, comments, etc... are welcome.

   If you send me a bug-report, please write your wget version, too.

   I don't use emacs-wget with Emacs/W3, so please give me an
   information about emacs-wget on Emacs/w3.


6. Acknowledgments

   This program shall never see the light without GNU wget, developer
   Hrvoje Niksic, and other wget contributors.  Thanks to great program.

   Great thanks to Takayuki Arakawa for his words, helps, and tests.
   Without him, it would take long time to develop this package.

   Juri Linkov <juri@jurta.org> greatly contributed to the development
   of emacs-wget 0.5.0.  He send me lots of ideas, suggestions, and
   bug-reports.  He also tested emacs-wget under development.

   Yoichi NAKAYAMA <yoichi@eken.phys.nagoya-u.ac.jp> maintains
   FreeBSD ports collection for emacs-wget.

   Akira TAGOH <tagoh@debian.org> maintains debian package for
   emacs-wget.

   Thanks to:

   Mitsugu SAKAMOTO <mitsugu@argv.org>
          - Reported unsupported functions on Meadow-pre1.15
          (window-text-height, replace-regexp-in-string).

   TSURUDA Naoki <tsuru@yokohama.office.ne.jp>
          - Reported bug running on Meadow-1.15.
          - Reported arguments order problem when using with wget-1.8.2 (cygwin).

   KOSEKI Yoshinori <kose@yk.NetLaputa.ne.jp>
          - Suggestion for load timing of w3m-wget.el.

   SHINYA Akihiko <shinya@sonicteam.com>
          - Reported problem when setting .wgetrc
          (quiet, timestamping, mirror, dir_prefix, and logfile).

   NAKAYA Toshiharu <nakaya8@is.titech.ac.jp>
          - Reported critical problem that emacs-wget removes LANG
          environment variable from Emacs' system variable.
          - Reported unsupported function on Emacs-20 or less
          (read-directory-name).

   Wataru Tachibana <MLB33828@nifty.com>
          - Suggestion for changing permission after downloading by its
          filename extension.

   mace <mace@kirjakaapeli.lib.hel.fi>
          - Suggestion for asking download directory always when user
          option wget-download-directory is nil.



Local Variables:
mode: indented-text
fill-column: 72
End:


debian/changelog
wget-el (0.5.0-11bullseye1) bullseye; urgency=low

  * Upstream updates.
  * Get rid of compilation warnings.
  * W3m compatibility fix.
  * Load customization options on startup.
  * Don't imply w3m loading on startup.
  * Packaged for Debian Bullseye.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Wed, 11 Jan 2023 06:55:04 +0300

wget-el (0.5.0-8) unstable; urgency=low

  * Use debian-pkg-add-load-path-item in startup script, as demanded by
    Debian Emacs policy. Also thanks to Kevin Ryde for spotting this
    (Closes: #582576).
  * Change email address to jmaslibre@debian.org.gt everywhere.
  * debian/wget-el.install added.
  * lpath.el removed because it is not needed (Closes: #582577).
  * debian/rules switch to dh 7.
  * debian/compat, raised to 7.
  * Switch to dpkg-source 3.0 (quilt) format.
  * debian/control:
    - Build dependen on version >= 7.0.50~ of debhelper because
      debhelper's overrides need it.
    - Updated policy to 3.9.1 (no changes needed)

 -- Josue Abarca <jmaslibre@debian.org.gt>  Tue, 27 Jul 2010 15:27:10 -0600

wget-el (0.5.0-7) unstable; urgency=low

  * debian/rules:
    - Removed unused option "dh_install -m 0664" (closes: Bug#518855)
    - dh_prep replaces "dh_clean -k" (deprecated)
  * Move debhelper compat level to 7
  * debian/control:
    - Updated policy to 3.8.1 (no changes needed)
    - The word "an" have been removed from the short description
    - Updated "debhelper" dependency

 -- Josue Abarca <jmaslibre@gmail.com>  Mon, 16 Mar 2009 00:17:04 -0600

wget-el (0.5.0-6) unstable; urgency=low

  * debian/wget-el.emacsen-startup: Use of eval-after-load replaced.
    (closes: Bug#492627)
  * debian/control:
    - updated policy to 3.8.0 (no changes needed)

 -- Josue Abarca <jmaslibre@gmail.com>  Wed, 20 Aug 2008 00:28:17 -0600

wget-el (0.5.0-5) unstable; urgency=low

  * debian/watch change to version 3, new sintaxys
    Thanks to Raphael Geissert <atomo64@gmail.com> (closes: Bug#450316)
  * New Maintainer
    Thanks to Akira Tagoh for his previus work (closes: Bug#435372)
  * move debhelper compat level to 6
  * debian/copyright  file format changed to be machine-interpretable
  * debian/rules:
    - used dh_install instead of manually installing files
    - removed unused: dh_installexamples, dh_installman, dh_link,
      dh_strip and dh_shlibdeps.
  * debian/control:
    - updated "debhelper" dependency
    - removed "shlibs:Depends"
    
 -- Josue Abarca <jmaslibre@gmail.com>  Tue, 15 Apr 2008 00:16:53 -0600

wget-el (0.5.0-4) unstable; urgency=low

  * Orphaning this package.

 -- Akira TAGOH <tagoh@debian.org>  Tue, 31 Jul 2007 18:19:23 +0900

wget-el (0.5.0-3) unstable; urgency=low

  * debian/wget-el.emacsen-startup: removed setq of wget-process-buffer.
    (closes: Bug#296028)

 -- Akira TAGOH <tagoh@debian.org>  Wed, 23 Feb 2005 20:34:09 +0900

wget-el (0.5.0-2) unstable; urgency=low

  * debian/wget-el.emacsen-startup: added quote.
    Thanks to Victor Cordero <victorctd@wanadoo.es>. (closes: Bug#284276)

 -- Akira TAGOH <tagoh@debian.org>  Tue,  7 Dec 2004 03:41:42 +0900

wget-el (0.5.0-1) unstable; urgency=low

  * New upstream release

 -- Akira TAGOH <tagoh@debian.org>  Tue, 26 Oct 2004 17:30:20 +0900

wget-el (0.4.2-3) unstable; urgency=low

  * wget-el.emacsen-install: gzip'ed CompilationLog with -f.
    (closes: Bug#277104)
  * wget-el.emacsen-startup: fixed a typo. Thanks to Aaron M. Ucko.

 -- Akira TAGOH <tagoh@debian.org>  Mon, 25 Oct 2004 21:07:30 +0900

wget-el (0.4.2-2) unstable; urgency=low

  * wget-el.emacsen-startup: set the variables after loading wget.el
    (closes: Bug#272154)

 -- Akira TAGOH <tagoh@debian.org>  Wed, 22 Sep 2004 09:09:23 +0900

wget-el (0.4.2-1) unstable; urgency=low

  * Initial Release. (closes: Bug#264724)

 -- Akira TAGOH <tagoh@debian.org>  Wed, 11 Aug 2004 14:53:05 +0900



w3m-wget.el
;;; w3m-wget.el --- Interface program of wget on emacs-w3m.

;; Copyright (C) 2001, 2002 Masayuki Ataka <masayuki.ataka@gmail.com>

;; Authors: Masayuki Ataka <masayuki.ataka@gmail.com>
;; Keywords: w3m, WWW, hypermedia

;; This file is a part of emacs-wget.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, you can either send email to this
;; program's maintainer or write to: The Free Software Foundation,
;; Inc.; 59 Temple Place, Suite 330; Boston, MA 02111-1307, USA.


;;; Code:
(require 'wget-custom)
(require 'w3m-util "w3m-util" 'noerror)

(autoload 'wget-api "wget" "Application Program Interface for wget")

(declare-function w3m-anchor "ext:w3m-util.el" (&optional position))
(declare-function w3m-image "ext:w3m-util.el" (&optional position))

(defvar w3m-current-title)
(defvar w3m-current-url)
(defvar w3m-mode-map)
(defvar wget-current-title)

(defun w3m-wget (arg)
  "Download anchor, image, or current page.
With prefix argument ARG, you can change uri."
  (interactive "P")
  (let ((uri (or (w3m-anchor) (w3m-image))))
    (setq wget-current-title w3m-current-title)
    (wget-api uri w3m-current-url arg)))

(when w3m-wget-substitute-download-command
  (substitute-key-definition 'w3m-download-this-url
                             'w3m-wget w3m-mode-map))


(provide 'w3m-wget)
;;; w3m-wget.el ends here


wget-custom.el
;;; wget-custom.el --- Customizable wget options.

;; Copyright (C) 2001, 2002 Masayuki Ataka <masayuki.ataka@gmail.com>

;; Authors: Masayuki Ataka <masayuki.ataka@gmail.com>
;; Keywords: w3m, WWW, hypermedia

;; This file is a part of emacs-wget.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, you can either send email to this
;; program's maintainer or write to: The Free Software Foundation,
;; Inc.; 59 Temple Place, Suite 330; Boston, MA 02111-1307, USA.


;;; Code:

(defgroup wget nil
  "wget program interface."
  :group 'hypermedia
  :prefix "wget-")


;;;; User Options

(defcustom wget-command "wget"
  "*Program name of `wget'."
  :group 'wget
  :type 'string)

(defcustom wget-default-options nil
  ;; '("-nc" "-c")                      ; <- example.
  "*List of default options of wget."
  :group 'wget
  :type '(repeat
	  (restricted-sexp :tag "Option")))

(defcustom wget-web-page-options
  `(,@wget-default-options
    "-r"				;--recursive: Turn on recursive retrieving.
    "-np")				;--no-parent: Do not ever ascend to the parent directory.
  "*List of options to download all Web pages."
  :group 'wget
  :type '(repeat
	  (restricted-sexp :tag "Web page Option")))

(defcustom wget-ftp-default-options nil
  "*List of default options when download from ftp.
If nil, use `wget-default-options' instead."
  :group 'wget
  :type '(repeat
	  (restricted-sexp :tag "FTP Option")))

;; Download dir

(defcustom wget-download-directory
  "~/download"
  "*Default directory name that retrieved files go.
If nil, always ask download directory."
  :group 'wget
  :type '(choice directory (sexp :tag "Lisp object")))

(defcustom wget-download-directory-filter nil
  "*Function that defines the filtering of download directory."
  :group 'wget
  :type '(radio (const :tag "No filter" nil)
		(const :tag "Filter by regexp"
		       wget-download-dir-filter-regexp)
		(const :tag "Aliases"
		       wget-download-dir-filter-alias)
		(const :tag "Filter by regexp and alias"
		       wget-download-dir-filter-regexp-and-alias)
		(const :tag "Check current dir"
		       wget-download-dir-filter-current-dir)
		(function :tag "Other Function")))

;; Download Log

(defcustom wget-download-log-file nil
  "*Default file name for download log.
If emacs-wget find this file in the download directory,
append download log in format `wget-download-log-format'.

If log file is not found, create or ask creating.
See the variable `wget-create-download-log'."
  :group 'wget
  :type 'file)

(defcustom wget-create-download-log 'always
  "*If non-nil, create or ask creating the log file
when the log file is not found in the download directory.

'always  Always create log file
'ask     Ask creating log file
nil      Do not create log file.

The log file name is defined by variable `wget-download-log-file'."
  :group 'wget
  :type '(choice (const always) (const ask) (const nil)))

(defcustom wget-download-log-format
  "%T\t%U\n"
  "*Format string for `wget-download-log-file'.

%T	Time format string replaced by `wget-download-log-time-format'
%t	Title
%U	URI"
  :group 'wget
  :type 'string)

(defcustom wget-download-log-time-format
  "%Y-%m-%d %H:%M:%S"
  "*Time format string for `wget-download-log-file'.
See the function `format-time-string' for format-string."
  :group 'wget
  :type 'string)

(defcustom wget-add-download-log-eof t
  "*If non-nil, download log is added at the end of file,
else it is added at the beginning of file.
Download log is added to the file `wget-download-log-file'
in the format `wget-download-log-format'."
  :group 'wget
  :type 'boolean)

;; misc

(defcustom wget-executable-file-extension-list nil
  ; '("exe" "sh" "csh" "pl" "rb") 	; <- example
  "*List of file extension that change file permission executable after downloading."
  :group 'wget
  :type '(repeat
	  (restricted-sexp :tag "File extension")))

(defcustom wget-truncate-partial-width-windows t
  "*Non-nil means truncate lines in *wget* buffer less than full frame wide."
  :group 'wget
  :type 'boolean)

(defcustom wget-max-window-height (/ (frame-height) 2)
  "*Max height of *wget* buffer."
  :group 'wget
  :type 'integer)

;; Hooks

(defcustom wget-hook nil
  "*Hook run after calling `wget-uri'."
  :group 'wget
  :type 'hook)

(defcustom wget-after-hook nil
  "*Hook run after finishing downloading file."
  :group 'wget
  :type 'hook)

(defcustom wget-load-hook nil
  "*Hook run after loading Emacs-wget."
  :group 'wget
  :type 'hook)

;; debug etc...

(defcustom wget-debug nil
  "*Non nil means save wget log message in buffer `wget-debug-buffer'."
  :group 'wget
  :type 'boolean)

;; w3m

(defgroup w3m-wget nil
  "wget interface for emacs-w3m."
  :group 'wget
  :group 'w3m
  :prefix "w3m-wget-")

(defcustom w3m-wget-substitute-download-command t
  "*If non-nil, substitute download command from emacs-w3m default to `w3m-wget'."
  :group 'w3m-wget
  :type 'boolean)


(provide 'wget-custom)
;;; wget-custom.el ends here
